staging/src/k8s.io/apiextensions-apiserver/test/integration/versioning_test.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/test/integration/testserver"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+)
+
+func TestVersionedNamspacedScopedCRD(t *testing.T) {
+	stopCh, apiExtensionClient, dynamicClient, err := testserver.StartDefaultServerWithClients()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer close(stopCh)
+
+	noxuDefinition := testserver.NewMultipleVersionNoxuCRD(apiextensionsv1beta1.NamespaceScoped)
+	err = testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, dynamicClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ns := "not-the-default"
+	testSimpleVersionedCRUD(t, ns, noxuDefinition, dynamicClient)
```

Comment:
I was expecting to see some of the following tests:
* a test that actually checked what got persisted in etcd matched the storage version
* a test that set up a CRD, persisted some data, changed the version, and made sure the previously persisted data could be read
* a test that verified discovery docs tracked the CRD through creation, version addition, version removal, and deletion

something like described in https://github.com/kubernetes/kubernetes/pull/63518#discussion_r187123115:

> definitely need a test that goes through a version lifecycle to flush out serving/storage issues, something like:
> 1. create with v1, ensure v1 appears in discovery, is usable, and persists an object as v1
> 2. add v2 with served=true, ensure v1 and v2 appear in discovery, are usable, and still persist as v1
> 3. set v2 stored=true, ensure new object persists as v2, update all existing objects and ensure they persist as v2
> 4. set v1 served=false, ensure v1 disappears from discovery and the api is no longer callable
> 5. remove v1 from status.storedVersions
> 6. remove v1 from versions

that planned as a followup?

Reply:
I suggest we do that as a follow up.

Reply:
See https://github.com/kubernetes/kubernetes/issues/64136 for followup details.